### ISSUE TITLE
A question about 'sample_stats' in a client.server_info() output

### ISSUE THREAD

**Issue by limash** [deepmind/reverb#27]
Hey,

I have some issues understanding a client.server_info() output regarding sample_stats.
For example, I run "Example 1: Overlapping Trajectories" from the [Reverb Tutorial](https://github.com/deepmind/reverb/blob/master/examples/demo.ipynb). Executing `client.server_info()` afterwards I got the output:

> {'my_table': TableInfo(name='my_table', sampler_options=prioritized {
>   priority_exponent: 0.8
> }
> , remover_options=fifo: true
> is_deterministic: true
> , max_size=1000000, max_times_sampled=0, rate_limiter_info=samples_per_insert: 1.0
> min_diff: -1.7976931348623157e+308
> max_diff: 1.7976931348623157e+308
> min_size_to_sample: 2
> insert_stats {
>   completed: 2
>   completed_wait_time {
>   }
>   pending_wait_time {
>   }
> }
> sample_stats {
>   completed: 64
>   completed_wait_time {
>   }
>   pending_wait_time {
>   }
> }
> , signature=None, current_size=2, num_episodes=1, num_deleted_episodes=0)}

It says that 64 items were sampled. But we sampled just once a batch of 2 items consisting of 3 timesteps each. What does the number 64 mean?

Also, could you please answer is it possible to disable reverb warning messages. For example, each sampling reverb notifies that

> [reverb/cc/client.cc:159] Sampler and server are owned by the same process so Table uniform_table is accessed directly 
> without gRPC.

which is sometimes inconvenient. 

Thank you!

**Comment by acassirer** [comment]
Hey Shamil,

The behaviour here is a bit tricky but what you are seeing is correct (ish) behaviour from incorrect usage of the dataset. The fact that you are seeing the log message multiple times is proof of this. That message is only printed when the c++ Sampler object is created which typically only happens once at the start of the process. However, if you were to create a **new iterator** for each sample then you would indeed see this message over and over again (so in some sense the annoying warning is actually doing its job by being annoying).

The second part in understanding why the number 64 is seen is a bit more tricky and involves some understanding of the internals. Maybe this example will help a bit:

```python
import time

def make_checker(client: reverb.Client):
  start_value = client.server_info()['my_table'].rate_limiter_info.sample_stats.completed

  def _checker(want_diff: int):
    new_value = client.server_info()['my_table'].rate_limiter_info.sample_stats.completed
    diff = new_value - start_value
    if diff != want_diff:
      raise AssertionError(f'{diff} != {want_diff}, {new_value} - {start_value} != {want_diff}')
  
  return _checker


checker = make_checker(client)
checker(0)

# Creating the iterator will trigger the creation of underlying c++ Sampler 
# object. This object will open a streaming connection to the server and start
# requesting samples. The default settings of the datasets just so happens to
# allow it to request 64 samples in one go.
iterator = iter(dataset)
time.sleep(0.5)  # Hacky way to allow the sampler to catch up.
checker(64)

# We can pull 31 batches (62 samples) from the dataset without trigger any new
# items to be sampled.
for _ in range(31):
  next(iterator)
time.sleep(0.5)  # Hacky way to allow the sampler to catch up.
checker(64)

# The 32th call will return the last 2 samples from the buffer. This will
# trigger 64 new samples to be pulled from the server.
next(iterator)
time.sleep(0.5)  # Hacky way to allow the sampler to catch up.
checker(128)
```

The final part of my answer is to thank you for highlighting two bugs!

The first being that the `max_in_flight_samples_per_worker` doesn't have any effect when the `LocalSampler` is used (related to the bypassing gRPC message). This was intentional but your problem really highlights the problem with this design. I'll make some changes to make it respect this value even for local sampling.

The second bug is that `flexible_batch_size` is completely ignored by the ReplayDataset (look at [__init__](https://github.com/deepmind/reverb/blob/master/reverb/dataset.py)). I found that by trying to show how the number 64 could be changed in the above example by setting `flexible_batch_size` to a value that is not 64. Seeing that it didn't actually worked highlighted the second bug.

I'll submit some changes addressing this issues.

Thanks for reporting your findings and sorry for the confusing it caused.

Cheers,
Albin

**Comment by acassirer** [comment]
Hey again,

The two bugs have now been fixed by [a0237d2](https://github.com/deepmind/reverb/commit/a0237d2be6869c6aa93a42d06a3aa87fea915bfb) and [e40c594](https://github.com/deepmind/reverb/commit/e40c594f85e7c93c903380e886abb2b4e631930a). The changes should roll out with the next nightly build but won't reach the stable for a while. It seems unlikely that these bugs should have any meaningful impact on an actual experiment but if you really need to this exact control before the next stable release then you could simply run the sampling from a separate process. This will enable the gRPC workers which do not suffer from these two bugs.

Cheers,
Albin

**Comment by limash** [comment]
Thank you for a comprehensive answer!

Cheers,
Shamil
